### DISCUSSION TITLE
How to reduce require time

### DISCUSSION THREAD

**Discussion by forthrin** [HoneyryderChuck/httpx#14] [Q&A] [answered]
`HTTPX` has a (relatively!) high footprint on startup compared to other libraries. How can this be reduced?

```
time   require
0.001  date
0.002  time
0.003  readline
0.004  fileutils
0.008  m3u8
0.012  yaml
0.025  nokogiri
0.028  httpx/connection
0.031  httpx/transcoder
0.033  httpx/pool
0.094  httpx
```

**Comment by HoneyryderChuck** [comment, marked as answer]
I think you're comparing apples to oranges. Most of the examples are from bundled/standard gems; if you know how `require` works, it should be obvious why they're so fast. Also, a lot of gems are known for lazy loading code (and `nokogiri` is mostly a C extension, not a lot of ruby code to read/parse/evaluate). 

If you want to make a fair comparison, you should put it against other http clients. and repeat your example a few times to take caches into account. And you'll see that, while there's substantial room for improvement in `httpx`, it's not that critical.

**Comment by forthrin** [comment]
Lazy loading is a good idea. The library shouldn't really need to do anything until you actually use it.

You may close the issue unless you want it as a reminder for future improvements.

**Reply by HoneyryderChuck** [reply]
Lazy loading is no free lunch either. For instance, you'll be working against CoW deployments if you do so. `httpx` also used to use autoload for a few things. It made requiring it quite unstable in multithreaded environments, because autoload is not thread-safe. I've also done a bit of introspection, and the outliers here are loading the openssl-related module (which works with stdlib `openssl`), and the HTTP/2 parser. So in the first case, the cost is diluted, as lots of other code will require `openssl`. In the last case, it's a feature other alternatives just do not provide.

I'll mark this as answered. Feel free to propose changes which improve the status quo that take the drawbacks into account.
